test: Add i2c emulation for targets

yperess opened the pull request:
Update i2c_emul.c to support i2c_target_register and i2c_target_unregister function calls as well as support address forwarding in emulation. Address forwarding helps us test IPCs in native sim. Instead of having to emulate 2 separate cores, we can forward read/write requests from one bus to another bus (effectively creating a loop). This was the same image can simulate both the controller and the target.

Commit:
test: Add i2c emulation for targets

Update i2c_emul.c to support i2c_target_register and i2c_target_unregister
function calls as well as support address forwarding in emulation.
Address forwarding helps us test IPCs in native sim. Instead of having to
emulate 2 separate cores, we can forward read/write requests from one bus
to another bus (effectively creating a loop). This way the same image can
simulate both the controller and the target.

Signed-off-by: Yuval Peress <peress@google.com>

keith-zephyr commented on doc/hardware/emulator/bus_emulators.rst:
```diff
@@ -144,6 +144,33 @@ Zephyr includes the following emulators:
+     #size-cells = <0>;
+     #forward-cells = <1>;
+     reg = <0x100 4>;
+     forwards = <&i2c1 0x20>;
```
Nit - it would be helpful to demonstrate forwarding 2 addresses to the same bus.

yperess replied:
Done

keith-zephyr commented on drivers/i2c/i2c_emul.c:
```diff
@@ -74,13 +83,96 @@ static int i2c_emul_get_config(const struct device *dev, uint32_t *dev_config)
+			for (uint32_t j = 0; j < msgs[i].len; ++j) {
+				int rc = 0;
+
+				if (j == 0) {
+					LOG_DBG("    Calling read_requested with data %p",
+						(void *)&msgs[i].buf[j]);
+					rc = callbacks->read_requested(data->target_cfg,
+								       &msgs[i].buf[j]);
+				} else {
+					LOG_DBG("    Calling read_processed with data %p",
+						(void *)&msgs[i].buf[j]);
+					rc = callbacks->read_processed(data->target_cfg,
+								       &msgs[i].buf[j]);
+				}
```
As a follow up can you add support for `CONFIG_I2C_TARGET_BUFFER_MODE`?

yperess replied:
Sure, added it to my TODO list

keith-zephyr commented on drivers/i2c/i2c_emul.c:
```diff
@@ -74,13 +83,96 @@ static int i2c_emul_get_config(const struct device *dev, uint32_t *dev_config)
+	struct i2c_emul_data *data = dev->data;
+
+	LOG_DBG("    has_target_cfg? %d", data->target_cfg != NULL);
+	if (data->target_cfg != NULL) {
```
Please add some comments that describe the 2 new use cases.  I2C transfers sent to an I2C target, and I2C transfers forwarded from a controller to target.

yperess replied:
Done

keith-zephyr commented on tests/drivers/i2c/i2c_emul/CMakeLists.txt:
```diff
@@ -0,0 +1,13 @@
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(i2c_emul)
+
+FILE(GLOB app_sources src/*.c)
```
`app_sources` isn't used.  Plus using GLOB in Cmake isn't recommended I believe.

yperess replied:
Thanks for catching

keith-zephyr commented on tests/drivers/i2c/i2c_emul/src/test_forwarding.cpp:
```diff
@@ -0,0 +1,202 @@
+	     "Forwarding misconfigured");
+
+/* Setup the fake functions for the I2C target. */
+FAKE_VALUE_FUNC(int, target_read_requested, struct i2c_target_config *, uint8_t *);
+FAKE_VALUE_FUNC(int, target_read_processed, struct i2c_target_config *, uint8_t *);
+FAKE_VALUE_FUNC(int, target_write_requested, struct i2c_target_config *);
+FAKE_VALUE_FUNC(int, target_write_received, struct i2c_target_config *, uint8_t);
+FAKE_VALUE_FUNC(int, target_stop, struct i2c_target_config *);
+
```
Can this move into the emulated_target.c file?
Or maybe just move the contents of emulated_target.c here.

yperess replied:
Sure, I had to turn emulated_target to a CPP file since the fakes are using lambdas, but it's much cleaner now.

keith-zephyr commented on tests/drivers/i2c/i2c_emul/src/emulated_target.c:
```diff
@@ -0,0 +1,20 @@
+
+struct i2c_target_config emulated_target_config = {
+	.flags = 0,
+	.address = 0x20,
```
Can you extract the address from the devicetree? 
What happens if someone updates the devicetree to add a second forwarded target?

yperess replied:
Done

keith-zephyr commented on tests/drivers/i2c/i2c_emul/src/test_forwarding.cpp:
```diff
@@ -0,0 +1,202 @@
+ZTEST_SUITE(i2c_emul_forwarding, NULL, i2c_emul_forwarding_setup, i2c_emul_forwarding_before, NULL,
+	    i2c_emul_forwarding_teardown);
+
+ZTEST(i2c_emul_forwarding, test_write_is_forwarded)
```
I think a `test_transfer_is_forwarded` would also be useful to validate a multi-stage transfer.

yperess replied:
Done, man, lambdas for the win on this.

aescolar reviewed (commented):
Thank you for updating the docs and adding a test.
No overall complaints from me from a quick pass.

There is a typo on the commit msg, given that anyhow you need to push again:
"This wa-s-+y+ the same image can simulate both the controller and the target."

Commit:
i2c: Add/update utility inline functions

- Change the argument for i2c_is_read_op to a const pointer
- Add function i2c_is_stop_op to test if a message has a stop flag.

Signed-off-by: Yuval Peress <peress@google.com>

Commit:
i2c_emul: Add support for CONFIG_I2C_TARGET_BUFFER_MODE

Add emulation and test to support the buffered target mode.

Signed-off-by: Yuval Peress <peress@google.com>

teburd commented on doc/hardware/emulator/bus_emulators.rst:
```diff
@@ -144,6 +144,33 @@ Zephyr includes the following emulators:
+same image.
+
+.. note::
+   The ``#forward-cells`` attribute should always be 1.
```
Can you elaborate on why it should always be 1 here, unclear to me

yperess replied:
Added more information using the example above.

teburd reviewed (commented):
I have a hard time reading the test scenarios in cpp, so I can't comment much on them other than... they are hard for me to read.

The requiement of the forwards = <1> is confusing and unclear, I think it needs some commentary.

teburd commented on tests/drivers/i2c/i2c_emul/src/test_forwarding_buf.cpp:
```diff
@@ -0,0 +1,148 @@
+{
+	uint8_t expected[] = {0x01, 0x02, 0x03};
+	uint8_t data[ARRAY_SIZE(expected)] = {};
+	target_buf_read_requested_0_fake.custom_fake = [&expected](struct i2c_target_config *,
```
I have no idea what ```[&expected](...)``` means here, this needs some commentary

yperess replied:
Added comments about lambdas & capture variables.

teburd commented on tests/drivers/i2c/i2c_emul/src/test_forwarding_buf.cpp:
```diff
@@ -0,0 +1,148 @@
+ZTEST(i2c_emul_forwarding, test_read_request_overflow)
+{
+	uint8_t data;
+	target_buf_read_requested_0_fake.custom_fake = [](struct i2c_target_config *, uint8_t **_,
```
similarly confusing syntax

teburd commented on tests/drivers/i2c/i2c_emul/src/test_forwarding_pio.cpp:
```diff
@@ -0,0 +1,254 @@
+	// Try reading some values
+	for (uint8_t i = 0; i < 10; ++i) {
+		const uint8_t expected_data[2] = {
+			static_cast<uint8_t>(0x1 * i),
```
can't we just cast? (uint8_t)(0x1*i), easy peasy

yperess replied:
I guess, but why not use the static_cast which verifies that the value can be cast statically? The main up side of the static_cast is that if we go back and change `i` or the constant we're multiplying by to ones that wouldn't be compatible with casting to a uint8, we would catch it at compile time. Overall the reason is just future proofing the cast.

teburd replied:
I don't really know what this means by future proofing here, a cast is a cast? The result type is uint8_t and well known here. Surprising to even see the need for a cast to be honest. Not a deal breaker for me, but was confusing to see.

teburd commented on tests/drivers/i2c/i2c_emul/src/test_forwarding_pio.cpp:
```diff
@@ -0,0 +1,254 @@
+		      "Expected to be called 1 time, got %d",
+		      target_read_processed_1_fake.call_count);
+
+	// Check that we didn't forward to the first target
```
Don't think we allow any comments besides /* */ in our style guide

yperess replied:
I checked https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html#main-rules and rule 3.2 explicitly mentions // comments though I'm not seeing any rules prohibiting them. Are there any specific objections to them in .cpp files? I 100% agree we can't use them in .h files and clearly .c

teburd replied:
I too couldn't find any rules about // comments after hunting around, though I very much recall a linter/styler checker/our guide saying something along those lines previously its clearly not bothering to check these .cpp files, and there's no definitive rules.

teburd commented on tests/drivers/i2c/i2c_emul/src/test_forwarding_pio.cpp:
```diff
@@ -0,0 +1,254 @@
+	zexpect_equal(0, target_stop_0_fake.call_count);
+}
+
+} // namespace
```
?

yperess replied:
I'm going to comment in the other thread since these comments are very related

teburd commented on tests/drivers/i2c/i2c_emul/src/test_forwarding_pio.cpp:
```diff
@@ -0,0 +1,254 @@
+#include <zephyr/fff.h>
+#include <zephyr/ztest.h>
+
+namespace
```
why an anonymous namespace here?

yperess replied:
Anonymous namespaces have a few advantages over declaring everything static. https://medium.com/pranayaggarwal25/unnamed-namespaces-static-f1498741c527 is a pretty good read. The closing comment is similar to the closing comment we put on long #ifdef so if we have nested namespaces it's easier to follow which one is closed.

teburd replied:
I feel like I'm going to have to learn more and more C++ to keep reviewing this, and I'd say this is a problem as I don't have any inclination, time, or energy to learn C++ in depth any time soon. This is problematic for me as the current maintainer of i2c. I think if closures are the key feature you are looking for from C++ sticking with that, and ensuring some solid comments around them is pretty crucial. Beyond that keeping C++ usage to a minimum would be my recommendation.

If I'm getting the gist of this though its an alternative to labeling variables/functions static to keep them private to the compilation unit. Except... the entire rest of the Zephyr code base uses this already. So why are we doing things different for one set of unit tests?

teburd replied:
Seems ok I suppose, not -1 worthy but I don't necessarily love this either.

teburd commented on tests/drivers/i2c/i2c_emul/src/test_fowarding_common.cpp:
```diff
@@ -0,0 +1,79 @@
+	zexpect_equal(0, target_write_received_0_fake.call_count);
+	zexpect_equal(0, target_buf_write_received_0_fake.call_count);
+	zexpect_equal(0, target_buf_read_requested_0_fake.call_count);
+    zexpect_equal(0, target_stop_0_fake.call_count);
```
Looks like some odd spacing in this function

yperess replied:
I'll take a look when I'm by a computer, I ran clang format on the file before uploading so I don't think it'll be different. I'll see if maybe a clang ignore is in order.

yperess replied:
You're right, I ran clang format then forgot to commit :)

teburd replied:
Thanks!

teburd reviewed (dismissed):
I don't see any rules here about // in the coding guidelines anymore, perhaps its no longer relevant. I don't disagree that in C++ // as a comment is incredibly commonplace.

Thanks for commenting and working through some of the nits

keith-zephyr commented on drivers/i2c/i2c_emul.c:
```diff
@@ -89,6 +90,41 @@ static int i2c_emul_send_to_target(const struct device *dev, struct i2c_msg *msg
+#ifdef CONFIG_I2C_TARGET_BUFFER_MODE
+	const struct i2c_emul_config *config = dev->config;
+
+	if (!config->target_pio_mode) {
```
This should be `target_buffer_mode`.  Programmed I/O usually means write/read one byte at time.

yperess replied:
I modeled this after the `i2c_ite_enchance` driver which uses the DT property `target-pio-mode` in order to default to the buffered mode.

keith-zephyr replied:
The `target-pio-mode` option is independent of the buffer mode API.  The buffer mode API is a property of the I2C target driver (such as the [eeprom_target](https://github.com/zephyrproject-rtos/zephyr/blob/af314643a32db3452dd77f20dee2807197b179cf/drivers/i2c/target/eeprom_target.c#L227)) and this really isn't a property of the I2C controller driver.

I do see that the `i2c_ite_enhance` driver does reference the `buf_write_received` and `buf_read_received` routines, but this is only done for DMA mode (`target-pio-mode = false`).

keith-zephyr commented on dts/bindings/i2c/zephyr,i2c-emul-controller.yaml:
```diff
@@ -10,6 +10,11 @@ include: i2c-controller.yaml
 properties:
   reg:
     required: true
+  target-pio-mode:
```
`target-buffer-mode`

yperess replied:
Marking resolved to keep the conversation in one thread. This is modeled after the `i2c_ite_enhance` driver which uses `target-pio-mode`.

keith-zephyr commented on dts/bindings/i2c/zephyr,i2c-emul-controller.yaml:
```diff
@@ -10,6 +10,11 @@ include: i2c-controller.yaml
+  target-pio-mode:
+    type: boolean
+    description: |
+      This option is used when the I2C target is enabled and it can support PIO
+      (Programmed I/O) mode for I2C target transfer.
```
```suggestion
      This option is used when the I2C target is enabled and it can support buffered
      mode for I2C target transfer.
```

yperess replied:
Leaving this unresolved just so I don't lose track of this change if we want to default to pio mode.

yperess commented on doc/hardware/emulator/bus_emulators.rst:
```diff
@@ -144,6 +144,38 @@ Zephyr includes the following emulators:
+In the binding of the I2C emulated bus, there's a custom property for address
+based forwarding. Given the following devicetree node:
+
+.. code-block::
```
Done

keith-zephyr commented on doc/hardware/emulator/bus_emulators.rst:
```diff
@@ -144,6 +144,38 @@ Zephyr includes the following emulators:
+
+.. note::
+   The ``#forward-cells`` attribute should always be 1. Each entry in the
+   ``fowards`` attribute consists of the phandle followed by the address. In
```
Typo:
```suggestion
   ``forwards`` attribute consists of the phandle followed by the address. In
```

yperess replied:
Done

keith-zephyr reviewed (commented):
LGTM - but there is still a typo in the documentation.
